File: qml/showdevant/showmanager.js
.pragma library

Qt.include("commands.js");

//var queryModel = Qt.createQmlObject("import com.guid75 1.0; SqlQueryModel { }", null);

function __parseShowEpisodes(showId, queryModel, error, root) {
	var seasonNum;
	var episodeNum;
	var season;
	var episode;
	var str;
	if (error || root.code !== 1) {
		console.log("Parsing error");
		return;
	}
	for (seasonNum in root.seasons) {
		if (root.seasons.hasOwnProperty(seasonNum)) {
			season = root.seasons[seasonNum];

			// store the season into db
			queryModel.query = "INSERT INTO season (show_id, number) VALUES ('%1', %2)".arg(showId).arg(season.number);
			//				tx.executeSql("INSERT INTO season (show_id, number) VALUES (?, ?)", [showId, season.number]);

			for (episodeNum in season.episodes) {
				if (season.episodes.hasOwnProperty(episodeNum)) {
					episode = season.episodes[episodeNum];
					// store the episode into db
					str = "INSERT INTO episode (show_id, season, episode, title, number, global, date, comments) " +
							"VALUES ('%1', %2, %3, '%4', '%5', %6, %7, %8)".arg(showId).
					arg(season.number).
					arg(episode.episode).
					arg(episode.title).
					arg(episode.number).
					arg(episode.global).arg(episode.date).arg(episode.comments);
					console.log(str);
//					queryModel.query = str;

/*					tx.executeSql("INSERT INTO episode (show_id, season, episode, title, number, global, date, comments) " +
								  "VALUES (?, ?, ?, ?, ?, ?, ?, ?)", [showId, season.number, episode.episode, episode.title, episode.number, episode.global, episode.date, episode.comments]);*/
				}
			}
		}
	}

	// update the episodes last check date of the show
/*	var now = new Date();
	tx.executeSql("UPDATE show SET episodes_last_check_date=%1 WHERE show_id='%2'".arg(now.getTime() / 1000).arg(showId));*/
}

function __fetchShow(showId) {
	showsEpisodes({ showId: showId }, function(error, root) {
		__parseShowEpisodes(showId, error, root);
	});

	//showsEpisodes({ showId: showId }, __parseShowEpisodes.bind(Qt, tx));
}

function fetchAllEpisodes(showId, queryModel, callback) {
	showsEpisodes({ showId: showId }, function(error, root) {
		__parseShowEpisodes(showId, queryModel, error, root);
		if (callback) {
			callback(!error);
		}
	});
}

// Load a show seasons and episodes
function load(showId, callback) {
	var lastCheckEpoch = 0;
	var expiration = 24 * 64 * 64 * 1000; // TODO make this value customizable
	var now;

	db.readTransaction(function(tx) {
		var expired = false;
		var rs = tx.executeSql('SELECT episodes_last_check_date FROM show WHERE show_id="%1"'.arg(showId));

		if (rs.rows.length > 0) {
			lastCheckEpoch = rs.rows.item(0).episodes_last_check_date * 1000;
		}

		now = new Date();
		if (now.getTime() - lastCheckEpoch > expiration) {
			expired = true;
			// expired data, we need to launch the request
			__fetchShow(showId);
		}
		callback(showId, expired);
	});
}

function addShow(showId, title, callback) {
	db.transaction(function(tx) {
		var ts = tx.executeSql("INSERT INTO show (show_id, title) VALUES (?, ?)", [showId, title]);
		var success = ts.rowsAffected > 0;
		if (success) {
			// we've got to pass a select to get the last inserted row
			// maybe there is a better way?
			ts = tx.executeSql("SELECT * from show WHERE rowid=?", [ts.insertId]);
			success = ts.rows.length === 1;
			if (success) {
				callback(success, ts.rows.item(0));
				return;
			}
		}
		callback(false);
	});
}

File: showmanager.cpp
//  Copyright 2013 Guillaume Denry (guillaume.denry@gmail.com)
//  This file is part of BetaSeeker.
//
//  BetaSeeker is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  BetaSeeker is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with BetaSeeker.  If not, see <http://www.gnu.org/licenses/>.

#include <QStringList>
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QDateTime>
#include <QVariant>
#include <QDebug>

#include "commandmanager.h"
#include "jsonparser.h"
#include "showmanager.h"

ShowManager *ShowManager::_instance = 0;

ShowManager &ShowManager::instance()
{
	if (!_instance)
		_instance = new ShowManager();
	return *_instance;
}

ShowManager::ShowManager() :
	QObject()
{
	connect(&CommandManager::instance(), &CommandManager::commandFinished,
			this, &ShowManager::commandFinished);
}

void ShowManager::commandFinished(int ticketId, const QByteArray &response)
{
	TicketData ticketData = parsing[ticketId];
	if (ticketData.showId.isNull())
		return;

	parsing.remove(ticketId);

	// call the parse function
	QMetaObject::invokeMethod(this, ticketData.parseMethodName.toLocal8Bit(), Q_ARG(QString, ticketData.showId), Q_ARG(QByteArray, response));

	emit refreshDone(ticketData.showId);
}

void ShowManager::parseEpisode(const QString &showId, int season, const QJsonObject &root)
{
	bool ok;
	int episode = root.value("episode").toString().toInt(&ok);
	if (!ok)
		return;
	QString title = root.value("title").toString();
	QString number = root.value("number").toString();
	int global = root.value("global").toString().toInt();
	int date = root.value("date").toDouble();
	int comments = root.value("comments").toString().toInt();

	QSqlQuery query;
	query.prepare("INSERT INTO episode (show_id, season, episode, title, number, global, date, comments) "
				  "VALUES (:show_id, :season, :episode, :title, :number, :global, :date, :comments)");
	query.bindValue(":show_id", showId);
	query.bindValue(":season", season);
	query.bindValue(":episode", episode);
	query.bindValue(":title", title);
	query.bindValue(":number", number);
	query.bindValue(":global", global);
	query.bindValue(":date", date);
	query.bindValue(":comments", comments);
	query.exec();
}

void ShowManager::parseEpisodes(const QString &showId, int season, const QJsonObject &root)
{
	QJsonObject episodesJson = root.value("episodes").toObject();
	foreach (const QString &key, episodesJson.keys()) {
		QJsonObject episodeJson = episodesJson.value(key).toObject();
		if (episodeJson.isEmpty())
			continue;
		parseEpisode(showId, season, episodeJson);
	}
}

void ShowManager::parseSeasons(const QString &showId, const QByteArray &response)
{
	JsonParser parser(response);
	if (!parser.isValid()) {
		// TODO manage error
		return;
	}

	QJsonObject seasonsJson = parser.root().value("seasons").toObject();
	if (!QSqlDatabase::database().transaction()) {
		qCritical("Error while beginning a transaction for SQL insertion");
		return;
	}

	foreach (const QString &key, seasonsJson.keys()) {
		QJsonObject seasonJson = seasonsJson.value(key).toObject();
		if (seasonJson.isEmpty())
			continue;

		// numbers are under the form of strings from betaseries :/
		bool ok;
		int number = seasonJson.value("number").toString().toInt(&ok);
		if (!ok)
			continue;

		QSqlQuery query;
		query.prepare("INSERT INTO season (show_id, number) "
					  "VALUES (:show_id, :number)");
		query.bindValue(":show_id", showId);
		query.bindValue(":number", number);
		query.exec();
		parseEpisodes(showId, number, seasonJson);
	}

	// update the episodes last check date of the show
	QSqlQuery query;
	query.prepare("UPDATE show SET episodes_last_check_date=:epoch WHERE show_id=:showid");
	query.bindValue(":epoch", QDateTime::currentMSecsSinceEpoch() / 1000);
	query.bindValue(":showid", showId);
	query.exec();
	QSqlDatabase::database().commit();
}

int ShowManager::refreshOnExpired(const QString &showid)
{
	QSqlQuery query;
	qint64 last_check_epoch = 0;
	qint64 expiration = 24 * 60 * 60 * 1000; // one day => TODO parametrable
	int ticket;
	// have we the season in database?
	// take the expiration date in account
	query.exec(QString("SELECT episodes_last_check_date FROM show WHERE show_id='%1'").arg(showid));
	if (query.next() && !query.value(0).isNull()) {
		last_check_epoch = query.value(0).toLongLong() * 1000;
	}
	if (QDateTime::currentDateTime().toMSecsSinceEpoch() - last_check_epoch > expiration) {
		// expired data, we need to launch the request
		ticket = CommandManager::instance().showsEpisodes(showid);
		// TODO ticket can be invalid, manage it
		parsing.insert(ticket, TicketData(showid, "parseSeasons"));
		return 1;
	} else
		return 0;
}
